amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java

```diff
@@ -1095,45 +1151,29 @@ public int compare(final Lease lease1, final Lease lease2) {
-                     * lease just like we do for TRIM_HORIZON. However we will only return back records with server-side
-                     * timestamp at or after the specified initial position timestamp.
-                     *
-                     * Shard structure (each level depicts a stream segment):
-                     * 0 1 2 3 4   5   - shards till epoch 102
-                     * \ / \ / |   |
-                     *  6   7  4   5   - shards from epoch 103 - 205
-                     *   \ /   |  /\
-                     *    8    4 9  10 - shards from epoch 206 (open - no ending sequenceNumber)
-                     *
-                     * Current leases: empty set
-                     *
-                     * For the above example, suppose the initial position in stream is set to AT_TIMESTAMP with
-                     * timestamp value 206. We will then create new leases for all the shards (with checkpoint set to
-                     * AT_TIMESTAMP), including the ancestor shards with epoch less than 206. However as we begin
-                     * processing the ancestor shards, their checkpoints would be updated to SHARD_END and their leases
-                     * would then be deleted since they won't have records with server-side timestamp at/after 206. And
-                     * after that we will begin processing the descendant shards with epoch at/after 206 and we will
-                     * return the records that meet the timestamp requirement for these shards.
-                     */
-                    if (isDescendant
-                            && !initialPosition.getInitialPositionInStream().equals(InitialPositionInStream.AT_TIMESTAMP)) {
-                        newLease.checkpoint(ExtendedSequenceNumber.TRIM_HORIZON);
-                    } else {
+                    // If shard is a descendant, the leases for its ancestors were already created above. Open shards
+                    // that are NOT descendants will not have been created yet, so we do so here. We will not create
+                    // leases for open shards that ARE descendants yet - leases for these shards will be created upon
+                    // SHARD_END of their parents.
+                    if (!isDescendant) {
+                        log.debug("{} : shardId {} has no ancestors. Creating a lease.", streamIdentifier, shardId);
+                        final Lease newLease = multiStreamArgs.isMultiStreamMode() ?
+                                newKCLMultiStreamLease(shard, multiStreamArgs.streamIdentifier()) :
+                                newKCLLease(shard);
                         newLease.checkpoint(convertToCheckpoint(initialPosition));
+                        log.debug("{} : Set checkpoint of {} to {}", streamIdentifier, newLease.leaseKey(), newLease.checkpoint());
+                        shardIdToNewLeaseMap.put(shardId, newLease);
                     }
```

Comment:
can we add a debug log for else case of being a descendant ?

Reply:
Sure, will add